### ISSUE TITLE
openjdk seems to escape the buildFHSUserEnv chroot and load a package from outside chroot at the expense of a chroot-local package.

### ISSUE THREAD

**Issue by spacekitteh** [NixOS/nixpkgs#21606]
Will update as I know more.

Relevant: https://github.com/NixOS/nixpkgs/issues/21222#issuecomment-269198363

https://github.com/google/conscrypt/issues/20

**Comment by spacekitteh** [comment]
[Here](https://android.googlesource.com/platform/external/conscrypt/+/android-7.1.1_r6/Android.mk#341)'s where the BoringSSL library is (or at least, should be) statically linked into the libconscrypt_openjdk_jni.so file, which should then be used instead of openjdk loading openssl from the host.

**Comment by spacekitteh** [comment]
Gonna tag peeps who know about the core infrastructure and buildFHSUserEnv.

@edolstra @fpletz @vcunat @joachifm @abbradar

**Comment by abbradar** [comment]
Can this be reproduced without building all the AOSP? I've successfully built and ran unit tests for Conscrypt (standalone) in a chrootenv. I don't remember anything specific which we do that can cause this (in my knowledge symbols can't be overridden with anything short of `LD_PRELOAD` -- and I haven't even imagined that _statically linked_ PIC code can be affected with this!).

**Comment by spacekitteh** [comment]
Yep. Try `make BasicDreams`

**Comment by abbradar** [comment]
Hm, that's better but I still don't have necessary space for downloading all the AOSP now (I can meddle with manifests but I don't have this much time).

Interesting fact: Arch Linux's Java doesn't reference `libcrypto`: ```
[root@e7e065246fe8 /]# ldd `which java`
	linux-vdso.so.1 (0x00007ffc5e96b000)
	libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007fc623d26000)
	libz.so.1 => /usr/lib/libz.so.1 (0x00007fc623b10000)
	libjli.so => /usr/lib/jvm/java-8-openjdk/jre/bin/../lib/amd64/jli/libjli.so (0x00007fc623902000)
	libdl.so.2 => /usr/lib/libdl.so.2 (0x00007fc6236fe000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007fc623360000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fc623f43000)
```

I think they link statically more things than we are.

**Comment by abbradar** [comment]
Oh, I think that's because we link to GTK+ in non-headless version.

Can you try to use `jre8_headless` in the chrootenv?

**Comment by spacekitteh** [comment]
Oh hey! Using jre8_headless works! <3

Now to figure out the underlying problem.

**Comment by abbradar** [comment]
I think that's because libcrypto loaded in the address space of Java takes priority over symbols in Conscrypt. I thought this was impossible; my theory is that GOT is filled from the same global namespace even for statically linked symbols when they are compiled as PIC. Strange! 
-- 
Nikolay.

**Comment by spacekitteh** [comment]
Yeah, that's what I was gonna explore as my third hypothesis, as it's a combination of the first two!

What do you think causes this? As in, which upstream?

**Comment by abbradar** [comment]
Depends on what do you perceive as a problem. I think the best course of action is to granularize our dynamic linking so that not the whole Java links to e.g. fontconfig or GTK but only the library that actually uses it. This will be a bit tedious but that's what Java library authors seem to expect. 

The root of the problem, if the theory is correct, needs design fixes for ld.so so that statically linked symbols can't be overridden like this. This is a bit more difficult to explore, implement and upstream ~_^. 
-- 
Nikolay.

**Comment by spacekitteh** [comment]
Weird to think we discovered a bug in such a commonly used component.

**Comment by spacekitteh** [comment]
Seems like it could be a pretty easy thing to exploit!

**Comment by abbradar** [comment]
I don't think it's a bug, rather it's a design choice (treat static and dynamic libraries as evenly as possible). That can be very evil however as we can see, especially in presence of inlining. That is, again, if the theory is correct. 
-- 
Nikolay.

**Comment by abbradar** [comment]
Hm, I don't see ways for this to be exploited any more than dynamic libraries - what do you mean? 
-- 
Nikolay.

**Comment by spacekitteh** [comment]
Hmm, I guess you're right; you'd need LD_PRELOAD or something similar.
